src/main/scala/mesosphere/marathon/storage/migration/MigrationTo18.scala

```diff
@@ -0,0 +1,175 @@
+
+  val migrationConditionReader = new Reads[Condition] {
+    private def readString(j: JsReadable) = j.validate[String].map {
+
+      case created if created.toLowerCase == "created" => Condition.Provisioned
+      case other => Condition(other)
+    }
+    override def reads(json: JsValue): JsResult[Condition] =
+      readString(json).orElse {
+        json.validate[JsObject].flatMap { obj => readString(obj \ "str") }
+      }
+  }
+
+  /**
+    * Read format for instance state where we replace created condition to provisioned.
+    */
+  val instanceStateReads17: Reads[InstanceState] = {
+    (
+      (__ \ "condition").read[Condition](migrationConditionReader) ~
+      (__ \ "since").read[Timestamp] ~
+      (__ \ "activeSince").readNullable[Timestamp] ~
+      (__ \ "healthy").readNullable[Boolean] ~
+      (__ \ "goal").read[Goal]
+    ) { (condition, since, activeSince, healthy, goal) =>
+        InstanceState(condition, since, activeSince, healthy, goal)
+      }
+  }
+
+  val taskStatusReads17: Reads[Task.Status] = {
+    (
+      (__ \ "stagedAt").read[Timestamp] ~
+      (__ \ "startedAt").readNullable[Timestamp] ~
+      (__ \ "mesosStatus").readNullable[MesosProtos.TaskStatus](Task.Status.MesosTaskStatusFormat) ~
+      (__ \ "condition").read[Condition](migrationConditionReader) ~
+      (__ \ "networkInfo").read[NetworkInfo](Formats.TaskStatusNetworkInfoFormat)
+
+    ) { (stagedAt, startedAt, mesosStatus, condition, networkInfo) =>
+        Task.Status(stagedAt, startedAt, mesosStatus, condition, networkInfo)
+      }
+  }
+
+  val taskReads17: Reads[Task] = {
+    (
+      (__ \ "taskId").read[Task.Id] ~
+      (__ \ "runSpecVersion").read[Timestamp] ~
+      (__ \ "status").read[Task.Status](taskStatusReads17)
+    ) { (taskId, runSpecVersion, status) =>
+        Task(taskId, runSpecVersion, status)
+      }
+  }
+
+  val taskMapReads17: Reads[Map[Task.Id, Task]] = {
+    mapReads(taskReads17).map {
+      _.map { case (k, v) => Task.Id(k) -> v }
+    }
+  }
+
+  /**
+    * Read format for old instance without goal.
+    */
+  val instanceJsonReads17: Reads[Instance] = {
+    (
+      (__ \ "instanceId").read[Instance.Id] ~
+      (__ \ "agentInfo").read[AgentInfo] ~
+      (__ \ "tasksMap").read[Map[Task.Id, Task]](taskMapReads17) ~
+      (__ \ "runSpecVersion").read[Timestamp] ~
+      (__ \ "state").read[InstanceState](instanceStateReads17) ~
+      (__ \ "unreachableStrategy").readNullable[raml.UnreachableStrategy] ~
+      (__ \ "reservation").readNullable[Reservation]
+    ) { (instanceId, agentInfo, tasksMap, runSpecVersion, state, ramlUnreachableStrategy, reservation) =>
+        val unreachableStrategy = ramlUnreachableStrategy.
+          map(Raml.fromRaml(_)).getOrElse(UnreachableStrategy.default())
+        new Instance(instanceId, Some(agentInfo), state, tasksMap, runSpecVersion, unreachableStrategy, reservation)
+      }
+  }
+
+  implicit val instanceResolver: IdResolver[Instance.Id, JsValue, String, ZkId] =
```

Comment:
Why do you need to define another `instanceResolver`? What's wrong with the existing `mesosphere/marathon/storage/store/ZkStoreSerialization.scala:63`?

Reply:
Check the return type: the one in ZkStoreSerialization returns `IdResolver[Instance.Id, Instance, String, ZkId]`, and I need a `JsValue`